jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/ServerWebSocketContainer.java

```diff
@@ -257,6 +265,55 @@ public boolean handle(Request request, Response response, Callback callback)
         if (negotiator == null)
             return false;
 
+        return upgrade(negotiator, request, response, callback);
+    }
+
+    /**
+     * <p>Upgrades the given {@code request} without matching against the WebSocket mappings.</p>
+     * <p>When {@code true} is returned, a response has been sent to the client
+     * and the {@code callback} has been completed; either because of a successful
+     * WebSocket upgrade, or because an error has occurred.</p>
+     * <p>When {@code false} is returned, a response has not been sent to the
+     * client, and the {@code callback} has not been completed; for example because
+     * the request is not a WebSocket upgrade; in this case the caller must arrange
+     * to send a response and complete the callback.</p>
+     *
+     * @param creator the creator of the WebSocket endpoint
+     * @param request the request to upgrade, possibly a WebSocket upgrade request
+     * @param response the response
+     * @param callback the callback to complete when the upgrade is complete
+     * @return {@code true} in case of WebSocket upgrades or failures,
+     * {@code false} if the request was not upgraded
+     * @see #handle(Request, Response, Callback)
+     */
+    public boolean upgrade(WebSocketCreator creator, Request request, Response response, Callback callback)
+    {
+        var coreCreator = newWebSocketCreator(creator);
+        WebSocketNegotiator negotiator = WebSocketNegotiator.from(coreCreator, factory, configuration);
+        return upgrade(negotiator, request, response, callback);
+    }
+
+    private org.eclipse.jetty.websocket.core.server.WebSocketCreator newWebSocketCreator(WebSocketCreator creator)
+    {
+        return (rq, rs, cb) ->
+        {
+            try
+            {
+                Object webSocket = creator.createWebSocket(new ServerUpgradeRequestDelegate(rq), new ServerUpgradeResponseDelegate(rq, rs), cb);
+                if (webSocket == null)
+                    cb.succeeded();
+                return webSocket;
+            }
+            catch (Throwable x)
+            {
+                cb.failed(x);
```

Comment:
Is it really correct to just fail the callback without doing any response.
Should this do some `Response.writeError()`?